File: web/src/lib/balance.ts
// Client port of fpledge.balance.check_balance — the structural squad-health flags. Kept
// faithful to the backend (same thresholds + messages) so it matches the server baseline
// on the recommended XI, then updates live as the user swaps bench <-> starters.

import type { SquadPlayer } from "@/lib/api";

const TEMPLATE_EO = 40.0; // ownership% at/above which a pick is "template"
const DIFFERENTIAL_EO = 10.0; // below which a pick is a "differential"
const ROTATION_MINUTES = 60.0; // expected minutes below which a starter is a rotation risk
const DEAD_BENCH_MINUTES = 20.0; // bench player below this won't provide auto-sub cover
const BENCH_SPEND_WARN = 18.0; // £m on the bench beyond which too much budget isn't scoring

/** A flag, plus the players it is actually about.
 *
 *  `message` stays byte-identical to the backend's — that parity is the point of this
 *  module. `detail` is additive and UI-only: a warning that says "3 starters are rotation
 *  risks" without naming them makes the reader go hunting through the pitch to find out
 *  who. Naming them is the difference between a verdict and something you can act on. */
export type BalanceFlag = {
  level: "warn" | "info" | "ok";
  message: string;
  detail?: { note: string; rows: { name: string; value: string }[] };
};

const round1 = (n: number) => Math.round(n * 10) / 10;

const byName = (a: SquadPlayer, b: SquadPlayer) => a.web_name.localeCompare(b.web_name);

export function computeHealth(squad: SquadPlayer[], starterIds: Set<number>): { flags: BalanceFlag[] } {
  const xi = squad.filter((p) => starterIds.has(p.element_id));
  const bench = squad.filter((p) => !starterIds.has(p.element_id));

  const benchSpend = round1(bench.reduce((s, p) => s + p.price, 0));

  const teamCounts = new Map<string, number>();
  for (const p of squad) teamCounts.set(p.team ?? "?", (teamCounts.get(p.team ?? "?") ?? 0) + 1);
  const topTwoClubs = [...teamCounts.values()].sort((a, b) => b - a).slice(0, 2).reduce((a, b) => a + b, 0);

  const xiXp = xi.reduce((s, p) => s + p.xp, 0);
  const nTemplate = xi.filter((p) => p.ownership >= TEMPLATE_EO).length;
  const nDifferential = xi.filter((p) => p.ownership < DIFFERENTIAL_EO).length;
  const nRotation = xi.filter((p) => (p.x_minutes ?? 90) < ROTATION_MINUTES).length;
  const deadBench = bench.filter((p) => (p.x_minutes ?? 90) < DEAD_BENCH_MINUTES).length;
  const top3 = [...xi.map((p) => p.xp)].sort((a, b) => b - a).slice(0, 3).reduce((a, b) => a + b, 0);
  const captainDependence = xiXp > 0 ? top3 / xiXp : 0;

  const flags: BalanceFlag[] = [];

  if (benchSpend > BENCH_SPEND_WARN)
    flags.push({
      level: "warn",
      message: `£${benchSpend.toFixed(1)}m on the bench — a lot of budget not scoring`,
      detail: {
        note: `Budget parked on players who only score if someone ahead of them doesn't play. Warns above £${BENCH_SPEND_WARN.toFixed(0)}m.`,
        rows: [...bench].sort((a, b) => b.price - a.price)
          .map((p) => ({ name: p.web_name, value: `£${p.price.toFixed(1)}m · ${p.xp.toFixed(1)} xP` })),
      },
    });

  if (deadBench >= 2)
    flags.push({
      level: "warn",
      message: `${deadBench} bench players unlikely to play — weak auto-sub cover`,
      detail: {
        note: `Under ${DEAD_BENCH_MINUTES} expected minutes, so they can't be relied on to cover a starter who blanks.`,
        rows: bench.filter((p) => (p.x_minutes ?? 90) < DEAD_BENCH_MINUTES).sort(byName)
          .map((p) => ({ name: p.web_name, value: `${Math.round(p.x_minutes ?? 0)}′ expected` })),
      },
    });

  if (topTwoClubs >= 6)
    flags.push({
      level: "warn",
      message: `${topTwoClubs}/${squad.length} players from just two clubs — concentration risk`,
      detail: {
        note: "One bad afternoon for either club takes a large share of your squad with it.",
        rows: [...teamCounts.entries()].sort((a, b) => b[1] - a[1]).slice(0, 2)
          .map(([club, n]) => ({
            name: club,
            value: `${n} · ${squad.filter((p) => (p.team ?? "?") === club).sort(byName).map((p) => p.web_name).join(", ")}`,
          })),
      },
    });

  if (nDifferential >= 5)
    flags.push({
      level: "warn",
      message: `${nDifferential} differentials in the XI — high variance`,
      detail: {
        note: `Owned by under ${DIFFERENTIAL_EO}% of managers. They win you rank when they return and cost you nothing when they blank — the swing cuts both ways.`,
        rows: xi.filter((p) => p.ownership < DIFFERENTIAL_EO).sort((a, b) => a.ownership - b.ownership)
          .map((p) => ({ name: p.web_name, value: `${p.ownership.toFixed(1)}% owned` })),
      },
    });

  if (nRotation >= 3)
    flags.push({
      level: "warn",
      message: `${nRotation} starters are rotation/minutes risks`,
      detail: {
        note: `Under ${ROTATION_MINUTES} expected minutes. Their xP is already reduced to match — this flags how much of your XI rests on it.`,
        rows: xi.filter((p) => (p.x_minutes ?? 90) < ROTATION_MINUTES).sort((a, b) => (a.x_minutes ?? 90) - (b.x_minutes ?? 90))
          .map((p) => ({ name: p.web_name, value: `${Math.round(p.x_minutes ?? 0)}′ expected` })),
      },
    });

  if (nTemplate >= 9)
    flags.push({
      level: "info",
      message: `very template XI (${nTemplate}/11 highly owned) — safe but low ceiling`,
      detail: {
        note: `Owned by ${TEMPLATE_EO}% or more. You move with the field rather than away from it: hard to fall behind, equally hard to climb.`,
        rows: xi.filter((p) => p.ownership >= TEMPLATE_EO).sort((a, b) => b.ownership - a.ownership)
          .map((p) => ({ name: p.web_name, value: `${p.ownership.toFixed(1)}% owned` })),
      },
    });

  if (captainDependence > 0.5)
    flags.push({
      level: "info",
      message: `${Math.round(captainDependence * 100)}% of XI xP from your top 3 — captain-dependent`,
      detail: {
        note: `${top3.toFixed(1)} of ${xiXp.toFixed(1)} projected points sit with three players. A quiet week from them is a quiet week overall.`,
        rows: [...xi].sort((a, b) => b.xp - a.xp).slice(0, 3)
          .map((p) => ({ name: p.web_name, value: `${p.xp.toFixed(1)} xP` })),
      },
    });

  if (!flags.some((f) => f.level === "warn"))
    flags.push({ level: "ok", message: "no structural warnings — squad looks balanced" });

  return { flags };
}

File: fpledge/balance.py
"""Squad balance check — a structural health report for a 15-man FPL squad.

Beyond raw xP: where the budget sits, whether the bench can actually cover (auto-subs),
club-concentration risk, template-vs-differential profile, rotation/availability risk, and
how captain-dependent the team is. Works on any squad — the optimiser's output or a real
FPL team pulled from the API.
"""

from __future__ import annotations

from collections import Counter
from collections.abc import Sequence

TEMPLATE_EO = 40.0        # ownership% at/above which a pick is "template"
DIFFERENTIAL_EO = 10.0    # ownership% below which a pick is a "differential"
ROTATION_MINUTES = 60.0   # expected minutes below which a starter is a rotation risk
DEAD_BENCH_MINUTES = 20.0 # bench player below this won't provide auto-sub cover
BENCH_SPEND_WARN = 18.0   # £m on the bench beyond which too much budget isn't scoring


def check_balance(players: Sequence[dict], budget: float = 100.0) -> dict:
    """Analyse a squad. Each player dict needs: name, position, price, team, xp, ownership,
    x_minutes, starter (bool), captain (bool). Returns metrics + human-readable flags.
    """
    xi = [p for p in players if p["starter"]]
    bench = [p for p in players if not p["starter"]]
    captain = next((p for p in players if p.get("captain")), None)

    total_cost = round(sum(p["price"] for p in players), 1)
    xi_xp = sum(p["xp"] for p in xi)
    total_xp = xi_xp + (captain["xp"] if captain else 0.0)  # captain counted twice

    spend_by_pos = {
        pos: round(sum(p["price"] for p in players if p["position"] == pos), 1)
        for pos in ("GK", "DEF", "MID", "FWD")
    }
    bench_spend = round(sum(p["price"] for p in bench), 1)

    team_counts = Counter(p["team"] for p in players)
    top_two_clubs = sum(c for _, c in team_counts.most_common(2))

    xi_ownership = sum(p["ownership"] for p in xi) / len(xi) if xi else 0.0
    n_template = sum(1 for p in xi if p["ownership"] >= TEMPLATE_EO)
    n_differential = sum(1 for p in xi if p["ownership"] < DIFFERENTIAL_EO)

    n_rotation = sum(1 for p in xi if p["x_minutes"] < ROTATION_MINUTES)
    dead_bench = sum(1 for p in bench if p["x_minutes"] < DEAD_BENCH_MINUTES)

    top3 = sum(sorted((p["xp"] for p in xi), reverse=True)[:3])
    captain_dependence = top3 / xi_xp if xi_xp > 0 else 0.0

    flags: list[tuple[str, str]] = []
    if bench_spend > BENCH_SPEND_WARN:
        flags.append(("warn", f"£{bench_spend}m on the bench — a lot of budget not scoring"))
    if dead_bench >= 2:
        flags.append(("warn", f"{dead_bench} bench players unlikely to play — weak auto-sub cover"))
    if top_two_clubs >= 6:
        flags.append((
            "warn",
            f"{top_two_clubs}/{len(players)} players from just two clubs — concentration risk",
        ))
    if n_differential >= 5:
        flags.append(("warn", f"{n_differential} differentials in the XI — high variance"))
    if n_rotation >= 3:
        flags.append(("warn", f"{n_rotation} starters are rotation/minutes risks"))
    if n_template >= 9:
        flags.append(("info", f"very template XI ({n_template}/11 highly owned) — safe but low ceiling"))
    if captain_dependence > 0.5:
        flags.append(("info", f"{captain_dependence:.0%} of XI xP from your top 3 — captain-dependent"))
    if not any(level == "warn" for level, _ in flags):
        flags.append(("ok", "no structural warnings — squad looks balanced"))

    return {
        "n_players": len(players),
        "total_cost": total_cost,
        "budget_left": round(budget - total_cost, 1),
        "total_xp": round(total_xp, 1),
        "xi_xp": round(xi_xp, 1),
        "value_pts_per_m": round(xi_xp / total_cost, 3) if total_cost else 0.0,
        "spend_by_pos": spend_by_pos,
        "bench_spend": bench_spend,
        "max_from_club": max(team_counts.values()),
        "distinct_clubs": len(team_counts),
        "xi_avg_ownership": round(xi_ownership, 1),
        "n_template": n_template,
        "n_differential": n_differential,
        "n_rotation_risk": n_rotation,
        "dead_bench": dead_bench,
        "captain_dependence": round(captain_dependence, 2),
        "flags": flags,
    }


def format_report(r: dict) -> str:
    """Render a balance report as a compact text block."""
    icon = {"warn": "⚠", "info": "•", "ok": "✓"}
    lines = [
        "\n=== SQUAD BALANCE ===",
        f"  cost £{r['total_cost']}m (bank £{r['budget_left']}m)  |  XI xP {r['xi_xp']}  |  "
        f"value {r['value_pts_per_m']} pts/£m",
        f"  spend  GK {r['spend_by_pos']['GK']}  DEF {r['spend_by_pos']['DEF']}  "
        f"MID {r['spend_by_pos']['MID']}  FWD {r['spend_by_pos']['FWD']}  (bench £{r['bench_spend']}m)",
        f"  clubs  {r['distinct_clubs']} distinct, max {r['max_from_club']} from one  |  "
        f"XI ownership avg {r['xi_avg_ownership']}%  "
        f"({r['n_template']} template, {r['n_differential']} differential)",
        f"  risk   {r['n_rotation_risk']} rotation-risk starters, {r['dead_bench']} dead bench, "
        f"captain-dependence {r['captain_dependence']:.0%}",
        "  flags:",
    ]
    lines += [f"    {icon.get(level, '-')} {msg}" for level, msg in r["flags"]]
    return "\n".join(lines)
